Add trusted files expansion in finalize_manifest.py

DukeDavis12 opened the pull request:
<!--
    Please fill in the following form before submitting this PR
    and ensure that your code follows our coding style guideline:
    https://gramine.readthedocs.io/en/latest/devel/coding-style.html -->

## Description of the changes <!-- (reasons and measures) -->
Fix gramine-direct failure while rendering entrypoint manifest file.

Commit aef087f "[LibOS] Move trusted and allowed files logic to LibOS"
in the core Gramine repository makes gramine-direct work with trusted
files, represented in TOML tables with each file's SHA256 hash and URI.

This commit updates finalize_manifest.py to incorporate trusted files
expansion/hashing logic, aligning the manifest with this change.

<!--
    If your PR fixes an issue, please remember to add "Fixes #issue_number"
    here, to automatically close it on merge. -->

## How to test this PR? <!-- (if applicable) -->
Manual Testing

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/gramineproject/gsc/230)
<!-- Reviewable:end -->

DukeDavis12 commented:
@kailun-qin  @woju @mkow  Please review.

woju reviewed (changes requested):
> Reviewed 1 of 1 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gsc/230)* status: all files reviewed, 3 unresolved discussions, not enough approvals from maintainers (3 more required), not enough approvals from different teams (2 more required, approved so far: ) (waiting on @DukeDavis12)

___
*[`finalize_manifest.py` line 36 at r1](https://reviewable.io/reviews/gramineproject/gsc/230#-OG__VX832rwcEki23Hp:-OG__VX832rwcEki23Hq:bckvg06) ([raw file](https://github.com/gramineproject/gsc/blob/5cef712c61bdee0c81cf628fe31020f3220c40a0/finalize_manifest.py#L36)):*
> ```Python
>     expanded_files = []
>     for uri in trusted_files:
>         file_path = urlparse(uri).path
> ```

Our `file:` are fake and do not conform to IANA `file:` urls. You should just `.split(':', 1)` them and check if the left part is `file'`, or equivalently `if uri.startswith('file:')`, or sth like that. In any case, please do not use `urllib` to parse them.
___
*[`finalize_manifest.py` line 37 at r1](https://reviewable.io/reviews/gramineproject/gsc/230#-OG__zzP8_jBebQwJGHX:-OG__zzP8_jBebQwJGHY:b-cxhm3z) ([raw file](https://github.com/gramineproject/gsc/blob/5cef712c61bdee0c81cf628fe31020f3220c40a0/finalize_manifest.py#L37)):*
> ```Python
>     for uri in trusted_files:
>         file_path = urlparse(uri).path
>         if os.path.exists(file_path):
> ```

And if it does not exist? What should be the expected behaviour if the file does not exist at the time of manifest generation?
___
*[`finalize_manifest.py` line 159 at r1](https://reviewable.io/reviews/gramineproject/gsc/230#-OG__LhX6orR7c8i3JjR:-OG__LhY3We1jz6OpsX5:bhyiwug) ([raw file](https://github.com/gramineproject/gsc/blob/5cef712c61bdee0c81cf628fe31020f3220c40a0/finalize_manifest.py#L159)):*
> ```Python
>     if 'allow_all_but_log' not in rendered_manifest_dict['sgx'].get('file_check_policy', ''):
>         trusted_files = generate_trusted_files(args.dir, already_added_files)
>         rendered_manifest_dict['sgx']['trusted_files'] = expand_trusted_files(trusted_files + already_added_files)
> ```

why did you change `.setdefault()` to ` = `?


<!-- Sent from Reviewable.io -->

DukeDavis12 commented:
> > Reviewed 1 of 1 files at r1, all commit messages.
> > _[Reviewable](https://reviewable.io/reviews/gramineproject/gsc/230)_ status: all files reviewed, 3 unresolved discussions, not enough approvals from maintainers (3 more required), not enough approvals from different teams (2 more required, approved so far: ) (waiting on @DukeDavis12)
> 
> _[`finalize_manifest.py` line 36 at r1](https://reviewable.io/reviews/gramineproject/gsc/230#-OG__VX832rwcEki23Hp:-OG__VX832rwcEki23Hq:bckvg06) ([raw file](https://github.com/gramineproject/gsc/blob/5cef712c61bdee0c81cf628fe31020f3220c40a0/finalize_manifest.py#L36)):_
> 
> > ```python
> >     expanded_files = []
> >     for uri in trusted_files:
> >         file_path = urlparse(uri).path
> > ```
> 
> Our `file:` are fake and do not conform to IANA `file:` urls. You should just `.split(':', 1)` them and check if the left part is `file'`, or equivalently `if uri.startswith('file:')`, or sth like that. In any case, please do not use `urllib` to parse them.
> 

- Updated.

> _[`finalize_manifest.py` line 37 at r1](https://reviewable.io/reviews/gramineproject/gsc/230#-OG__zzP8_jBebQwJGHX:-OG__zzP8_jBebQwJGHY:b-cxhm3z) ([raw file](https://github.com/gramineproject/gsc/blob/5cef712c61bdee0c81cf628fe31020f3220c40a0/finalize_manifest.py#L37)):_
> 
> > ```python
> >     for uri in trusted_files:
> >         file_path = urlparse(uri).path
> >         if os.path.exists(file_path):
> > ```
> 
> And if it does not exist? What should be the expected behaviour if the file does not exist at the time of manifest generation?
> 

-  Updated.

> _[`finalize_manifest.py` line 159 at r1](https://reviewable.io/reviews/gramineproject/gsc/230#-OG__LhX6orR7c8i3JjR:-OG__LhY3We1jz6OpsX5:bhyiwug) ([raw file](https://github.com/gramineproject/gsc/blob/5cef712c61bdee0c81cf628fe31020f3220c40a0/finalize_manifest.py#L159)):_
> 
> > ```python
> >     if 'allow_all_but_log' not in rendered_manifest_dict['sgx'].get('file_check_policy', ''):
> >         trusted_files = generate_trusted_files(args.dir, already_added_files)
> >         rendered_manifest_dict['sgx']['trusted_files'] = expand_trusted_files(trusted_files + already_added_files)
> > ```
> 
> why did you change `.setdefault()` to `=`?

 .setdefault() is designed to only set a value if the key does not already exist in the dictionary.
 
 The above approach ensures that the trusted_files key in the rendered_manifest_dict['sgx'] dictionary is set to the result of expand_trusted_files(trusted_files + already_added_files), regardless of whether the key was previously present or not (Creates key if not present).
 
 However, if you want to use .setdefault() to ensure the key exists and then update its value, you can do it in two steps:
1. Use .setdefault() to ensure the key exists.
2. Update the value of the key.

I have updated PR to reflect the last method of using setDefault and then updating the value.

The reason why I am not extending trusted_file is that the trusted_files might already contain entry in trusted_files= [ "URI", "URI"] format.
Example: https://github.com/gramineproject/gsc/blob/master/test/generic.manifest#L5 ; extending just appends expanded_files to trusted_file along with existing ["URI", "URI"]  entry resulting in failure during run-time.

kailun-qin reviewed (changes requested):
> Reviewed 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gsc/230)* status: all files reviewed, 7 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (2 more required, approved so far: ), one-liner is not in imperative mood (waiting on @woju)

___
*[`finalize_manifest.py` line 33 at r2](https://reviewable.io/reviews/gramineproject/gsc/230#-OKb0vKy7vd0lmfLc5VM:-OKb0vKz2-pseRpRpz8X:b-u56bpr) ([raw file](https://github.com/gramineproject/gsc/blob/2f7e99d262b37a9502a88fd87ae3e3f2acac6c7f/finalize_manifest.py#L33)):*
> ```Python
> def compute_sha256(filename):
>     sha256 = hashlib.sha256()
>     with open(filename, "rb") as f:
> ```

For consistency, pls use single quotes.



_Code quote:_
```Python
"rb"
```
___
*[`finalize_manifest.py` line 34 at r2](https://reviewable.io/reviews/gramineproject/gsc/230#-OKb121RBELxco-S1l2l:-OKb121RBELxco-S1l2m:b-7ropln) ([raw file](https://github.com/gramineproject/gsc/blob/2f7e99d262b37a9502a88fd87ae3e3f2acac6c7f/finalize_manifest.py#L34)):*
> ```Python
>     sha256 = hashlib.sha256()
>     with open(filename, "rb") as f:
>         for byte_block in iter(lambda: f.read(128 * sha256.block_size), b""):
> ```

ditto



_Code quote:_
```Python
""
```
___
*[`finalize_manifest.py` line 42 at r2](https://reviewable.io/reviews/gramineproject/gsc/230#-OKb1AS79o-3kYyB2LtL:-OKb1AS79o-3kYyB2LtM:b1skr8v) ([raw file](https://github.com/gramineproject/gsc/blob/2f7e99d262b37a9502a88fd87ae3e3f2acac6c7f/finalize_manifest.py#L42)):*
> ```Python
>     for uri in trusted_files:
>         file_path = uri2path(uri)
>         if os.path.exists(file_path):
> ```

Can't we use `file_path.exists()` here?



_Code quote:_
```Python
os.path.exists
```
___
*[`finalize_manifest.py` line 43 at r2](https://reviewable.io/reviews/gramineproject/gsc/230#-OKb16Fm0m8vvBGxqaNQ:-OKb16Fm0m8vvBGxqaNR:b-7ropln) ([raw file](https://github.com/gramineproject/gsc/blob/2f7e99d262b37a9502a88fd87ae3e3f2acac6c7f/finalize_manifest.py#L43)):*
> ```Python
>         file_path = uri2path(uri)
>         if os.path.exists(file_path):
>             expanded_files.append({"uri": uri,"sha256": compute_sha256(file_path)})
> ```

ditto



_Code quote:_
```Python
expanded_files.append({"uri": uri,"sha256": compute_sha256(file_path)})
```


<!-- Sent from Reviewable.io -->

kailun-qin reviewed (commented):
@DukeDavis12: Can you pls reply to my comments so that I can resolve them? Thanks!

> *[Reviewable](https://reviewable.io/reviews/gramineproject/gsc/230#-:-OLdgXRq6oML2N-lABn_:bixrl6q)* status: 0 of 1 files reviewed, 7 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (2 more required, approved so far: ), one-liner is not in imperative mood (waiting on @DukeDavis12 and @woju)



<!-- Sent from Reviewable.io -->

DukeDavis12 commented:
> > Reviewed 1 of 1 files at r2, all commit messages.
> > _[Reviewable](https://reviewable.io/reviews/gramineproject/gsc/230)_ status: all files reviewed, 7 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (2 more required, approved so far: ), one-liner is not in imperative mood (waiting on @woju)
> 
> _[`finalize_manifest.py` line 33 at r2](https://reviewable.io/reviews/gramineproject/gsc/230#-OKb0vKy7vd0lmfLc5VM:-OKb0vKz2-pseRpRpz8X:b-u56bpr) ([raw file](https://github.com/gramineproject/gsc/blob/2f7e99d262b37a9502a88fd87ae3e3f2acac6c7f/finalize_manifest.py#L33)):_
> 
> > ```python
> > def compute_sha256(filename):
> >     sha256 = hashlib.sha256()
> >     with open(filename, "rb") as f:
> > ```
> 
> For consistency, pls use single quotes.

Updated.

> 
> _Code quote:_
> 
> ```python
> "rb"
> ```
> 
> _[`finalize_manifest.py` line 34 at r2](https://reviewable.io/reviews/gramineproject/gsc/230#-OKb121RBELxco-S1l2l:-OKb121RBELxco-S1l2m:b-7ropln) ([raw file](https://github.com/gramineproject/gsc/blob/2f7e99d262b37a9502a88fd87ae3e3f2acac6c7f/finalize_manifest.py#L34)):_
> 
> > ```python
> >     sha256 = hashlib.sha256()
> >     with open(filename, "rb") as f:
> >         for byte_block in iter(lambda: f.read(128 * sha256.block_size), b""):
> > ```
> 
> ditto

Updated.
> 
> _Code quote:_
> 
> ```python
> ""
> ```
> 
> _[`finalize_manifest.py` line 42 at r2](https://reviewable.io/reviews/gramineproject/gsc/230#-OKb1AS79o-3kYyB2LtL:-OKb1AS79o-3kYyB2LtM:b1skr8v) ([raw file](https://github.com/gramineproject/gsc/blob/2f7e99d262b37a9502a88fd87ae3e3f2acac6c7f/finalize_manifest.py#L42)):_
> 
> > ```python
> >     for uri in trusted_files:
> >         file_path = uri2path(uri)
> >         if os.path.exists(file_path):
> > ```
> 
> Can't we use `file_path.exists()` here?

Updated with file_path.exists()
> 
> _Code quote:_
> 
> ```python
> os.path.exists
> ```
> 
> _[`finalize_manifest.py` line 43 at r2](https://reviewable.io/reviews/gramineproject/gsc/230#-OKb16Fm0m8vvBGxqaNQ:-OKb16Fm0m8vvBGxqaNR:b-7ropln) ([raw file](https://github.com/gramineproject/gsc/blob/2f7e99d262b37a9502a88fd87ae3e3f2acac6c7f/finalize_manifest.py#L43)):_
> 
> > ```python
> >         file_path = uri2path(uri)
> >         if os.path.exists(file_path):
> >             expanded_files.append({"uri": uri,"sha256": compute_sha256(file_path)})
> > ```
> 
> ditto

Updated.
> 
> _Code quote:_
> 
> ```python
> expanded_files.append({"uri": uri,"sha256": compute_sha256(file_path)})
> ```

DukeDavis12 reviewed (commented):
Resolved your comments.

> *[Reviewable](https://reviewable.io/reviews/gramineproject/gsc/230#-:-OM5u7JBBOnIWa1qcdwr:b-kxfk40)* status: 0 of 1 files reviewed, 7 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (2 more required, approved so far: ), one-liner is not in imperative mood (waiting on @kailun-qin and @woju)



<!-- Sent from Reviewable.io -->

woju reviewed (changes requested):
> Reviewed 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gsc/230)* status: all files reviewed, 5 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (2 more required, approved so far: ), one-liner is not in imperative mood (waiting on @DukeDavis12 and @kailun-qin)

___
*[`finalize_manifest.py` line 159 at r1](https://reviewable.io/reviews/gramineproject/gsc/230#-OG__LhX6orR7c8i3JjR:-OMEOVGk32fCv1jRpJvf:b1z1ewt) ([raw file](https://github.com/gramineproject/gsc/blob/5cef712c61bdee0c81cf628fe31020f3220c40a0/finalize_manifest.py#L159)):*
> The reason why I am not extending `trusted_files` is that the `trusted_files` might already contain entry in `trusted_files= [ "URI", "URI"]` format.

Thanks, this ^ was the explanation I was missing.

Adding separate `.setdefault()` doesn't make sense, please leave only `=` line.


<!-- Sent from Reviewable.io -->

DukeDavis12 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gsc/230)* status: 0 of 1 files reviewed, 5 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (2 more required, approved so far: ), one-liner is not in imperative mood (waiting on @kailun-qin and @woju)

___
*[`finalize_manifest.py` line 159 at r1](https://reviewable.io/reviews/gramineproject/gsc/230#-OG__LhX6orR7c8i3JjR:-OMFrwUL-WDaWRTh2gea:b-snyxcn) ([raw file](https://github.com/gramineproject/gsc/blob/5cef712c61bdee0c81cf628fe31020f3220c40a0/finalize_manifest.py#L159)):*
<details><summary><i>Previously, woju (Wojtek Porczyk) wrote…</i></summary><blockquote>

> The reason why I am not extending `trusted_files` is that the `trusted_files` might already contain entry in `trusted_files= [ "URI", "URI"]` format.

Thanks, this ^ was the explanation I was missing.

Adding separate `.setdefault()` doesn't make sense, please leave only `=` line.
</blockquote></details>

Removed setDefault().


<!-- Sent from Reviewable.io -->

kailun-qin reviewed (approved):
> Reviewed 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gsc/230)* status: all files reviewed, all discussions resolved, one-liner is not in imperative mood



<!-- Sent from Reviewable.io -->

kailun-qin reviewed (changes requested):
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gsc/230)* status: all files reviewed, 4 unresolved discussions, not enough approvals from maintainers (1 more required), not enough approvals from different teams (1 more required, approved so far: ITL)

___
*[`-- commits` line 2 at r5](https://reviewable.io/reviews/gramineproject/gsc/230#-OMPMtTV2iy4wpjOw4VZ:-OMPMtTV2iy4wpjOw4V_:burjsf8):*
What does this mean? `gramine-direct` does not render manifests from templates.



_Code quote:_
```
Fix gramine-direct failure on rendering manifest
```
___
*[`-- commits` line 4 at r5](https://reviewable.io/reviews/gramineproject/gsc/230#-OMPIQuZE-dCpTPVQk9c:-OMPIQuZE-dCpTPVQk9d:b-7ropln):*
ditto



_Code quote:_
```
 Fix gramine-direct failure on rendering entrypoint manifest file.
```
___
*[`-- commits` line 6 at r5](https://reviewable.io/reviews/gramineproject/gsc/230#-OMPMIFc6pDnxjRTw_Pe:-OMPMIFc6pDnxjRTw_Pf:b-pxnduu):*
This sentence seems incomplete, shouldn't it provide contrast like "expects ..., but ..." to describe the cause?



_Code quote:_
```
Gramine-direct expects trusted files to be in a TOML table structure.
```
___
*[`-- commits` line 8 at r5](https://reviewable.io/reviews/gramineproject/gsc/230#-OMPLaqY67YDyiPHaIzw:-OMPLaqY67YDyiPHaIzx:bhoch7z):*
Pls double check -- isn't this commit more relevant: https://github.com/gramineproject/gramine/pull/1820/commits/72668bb391d9a311148d8e824fb006f314e8dbba?



_Code quote:_
```
  Changes based on commit aef087f "[LibOS] Move trusted and allowed files
  logic to LibOS".
```


<!-- Sent from Reviewable.io -->

DukeDavis12 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gsc/230)* status: all files reviewed, 4 unresolved discussions, not enough approvals from maintainers (1 more required), not enough approvals from different teams (1 more required, approved so far: ITL) (waiting on @kailun-qin)

___
*[`-- commits` line 2 at r5](https://reviewable.io/reviews/gramineproject/gsc/230#-OMPMtTV2iy4wpjOw4VZ:-OMQ77oa1SUhPb2GYytz:bgav17o):*
<details><summary><i>Previously, kailun-qin (Kailun Qin) wrote…</i></summary><blockquote>

What does this mean? `gramine-direct` does not render manifests from templates.
</blockquote></details>

Updated commit header and message.
___
*[`-- commits` line 4 at r5](https://reviewable.io/reviews/gramineproject/gsc/230#-OMPIQuZE-dCpTPVQk9c:-OMQ7DT793AEiaIzH5gD:bgav17o):*
<details><summary><i>Previously, kailun-qin (Kailun Qin) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

Updated commit header and message.
___
*[`-- commits` line 6 at r5](https://reviewable.io/reviews/gramineproject/gsc/230#-OMPMIFc6pDnxjRTw_Pe:-OMQ7GAC2ALsHlhpP_z4:bgav17o):*
<details><summary><i>Previously, kailun-qin (Kailun Qin) wrote…</i></summary><blockquote>

This sentence seems incomplete, shouldn't it provide contrast like "expects ..., but ..." to describe the cause?
</blockquote></details>

Updated commit header and message.
___
*[`-- commits` line 8 at r5](https://reviewable.io/reviews/gramineproject/gsc/230#-OMPLaqY67YDyiPHaIzw:-OMQ7IMg75DoYqggm6GX:b7h37wf):*
<details><summary><i>Previously, kailun-qin (Kailun Qin) wrote…</i></summary><blockquote>

Pls double check -- isn't this commit more relevant: https://github.com/gramineproject/gramine/pull/1820/commits/72668bb391d9a311148d8e824fb006f314e8dbba?
</blockquote></details>

In gramine-direct mode startup; the application fails at point where commit aef08f changes are made.

https://github.com/gramineproject/gramine/blob/master/libos/src/fs/chroot/trusted.c#L408


<!-- Sent from Reviewable.io -->

kailun-qin reviewed (changes requested):
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gsc/230)* status: all files reviewed, 2 unresolved discussions, not enough approvals from maintainers (1 more required), not enough approvals from different teams (1 more required, approved so far: ITL)

___
*[`-- commits` line 8 at r5](https://reviewable.io/reviews/gramineproject/gsc/230#-OMPLaqY67YDyiPHaIzw:-OMQXorl8lBNIL3IaZS-:blb4i8a):*
<details><summary><i>Previously, DukeDavis12 (Davis Benny) wrote…</i></summary><blockquote>

In gramine-direct mode startup; the application fails at point where commit aef08f changes are made.

https://github.com/gramineproject/gramine/blob/master/libos/src/fs/chroot/trusted.c#L408
</blockquote></details>

Ok, I see (I mixed w/ the other issue that we just fixed), then pls do NOT include the commit I mentioned above.
___
*[`-- commits` line 2 at r6](https://reviewable.io/reviews/gramineproject/gsc/230#-OMQZ2_EDOV81t66KyYA:-OMQZ2_F4XMYtj8rKijZ:b27x6yg):*
-> Add trusted files expansion in `finalize_manifest.py`



_Code quote:_
```
Add SHA256 computation and expand trusted files
```
___
*[`-- commits` line 10 at r6](https://reviewable.io/reviews/gramineproject/gsc/230#-OMQbO3nA_S7S0cNT0uy:-OMQbO3oCVmYpSotNw6r:b-5dq71b):*
I suggest sth. like below:

Commit aef087f "[LibOS] Move trusted and allowed files logic to LibOS" in the core Gramine repository makes `gramine-direct` work with trusted files, represented in TOML tables with each file's SHA256 hash and URI.

This commit updates `finalize_manifest.py` to incorporate trusted files expansion/hashing logic, aligning the manifest with this change.



_Code quote:_
```
  - Adds a `compute_sha256` function to calculate the SHA256 hash of files.
  - Implements `expand_trusted_files` to expand trusted file URIs into a
  list of dictionaries containing the file URI and its SHA256 hash.
  - Fixes issue while starting container in gramine-direct mode.

  Changes based on commit aef087f "[LibOS] Move trusted and allowed files
  logic to LibOS" and 72668bb "[PAL] Validate entrypoint ELF file separately".
```


<!-- Sent from Reviewable.io -->

DukeDavis12 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gsc/230)* status: all files reviewed, 2 unresolved discussions, not enough approvals from maintainers (1 more required), not enough approvals from different teams (1 more required, approved so far: ITL) (waiting on @kailun-qin)

___
*[`-- commits` line 2 at r6](https://reviewable.io/reviews/gramineproject/gsc/230#-OMQZ2_EDOV81t66KyYA:-OMiKIOlEfNKrJCt3WfJ:b-bw8e14):*
<details><summary><i>Previously, kailun-qin (Kailun Qin) wrote…</i></summary><blockquote>

-> Add trusted files expansion in `finalize_manifest.py`
</blockquote></details>

Updated.
___
*[`-- commits` line 10 at r6](https://reviewable.io/reviews/gramineproject/gsc/230#-OMQbO3nA_S7S0cNT0uy:-OMiKEN25CSZApxIiXa5:b-bw8e14):*
<details><summary><i>Previously, kailun-qin (Kailun Qin) wrote…</i></summary><blockquote>

I suggest sth. like below:

Commit aef087f "[LibOS] Move trusted and allowed files logic to LibOS" in the core Gramine repository makes `gramine-direct` work with trusted files, represented in TOML tables with each file's SHA256 hash and URI.

This commit updates `finalize_manifest.py` to incorporate trusted files expansion/hashing logic, aligning the manifest with this change.
</blockquote></details>

Updated.


<!-- Sent from Reviewable.io -->

kailun-qin reviewed (approved):
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gsc/230)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved



<!-- Sent from Reviewable.io -->
